asserts/validation_set_test.go

```diff
@@ -167,6 +172,41 @@ func (vss *validationSetSuite) TestSnapRevisionOptional(c *C) {
 	c.Check(snaps[0].Revision, Equals, 0)
 }
 
+func (vss *validationSetSuite) TestSnapComponents(c *C) {
+	encoded := strings.Replace(validationSetExample, "TSLINE", vss.tsLine, 1)
+
+	const components = `    components:
+      string-only: optional
+      with-revision:
+        revision: 10
+        presence: required
+      no-revision:
+        presence: invalid
+`
+
+	encoded = strings.Replace(encoded, "OTHER", components, 1)
+
+	a, err := asserts.Decode([]byte(encoded))
+	c.Assert(err, IsNil)
+	c.Check(a.Type(), Equals, asserts.ValidationSetType)
+
+	valset := a.(*asserts.ValidationSet)
+	snaps := valset.Snaps()
+	c.Assert(snaps, HasLen, 1)
+	c.Assert(snaps[0].Components, HasLen, 3)
+
+	c.Check(snaps[0].Components["string-only"], Equals, asserts.ValidationSetComponent{
+		Presence: asserts.PresenceOptional,
+	})
+	c.Check(snaps[0].Components["with-revision"], Equals, asserts.ValidationSetComponent{
+		Presence: asserts.PresenceRequired,
+		Revision: 10,
+	})
+	c.Check(snaps[0].Components["no-revision"], Equals, asserts.ValidationSetComponent{
+		Presence: asserts.PresenceInvalid,
+	})
```

Comment:
We could join these checks
```suggestion
	c.Assert(snaps[0].Components, DeepEquals, map[string]asserts.ValidationSetComponent{
		"string-only": asserts.ValidationSetComponent{
			Presence: asserts.PresenceOptional,
		},
		"with-revision": asserts.ValidationSetComponent{
			Presence: asserts.PresenceRequired,
			Revision: 10,
		},
		"no-revision": asserts.ValidationSetComponent{
			Presence: asserts.PresenceInvalid,
		},
	})
```